src/SAWScript/Yosys/CompositionalTranslation.hs

```diff
@@ -0,0 +1,343 @@
+      Nothing -> case c ^. cellType of
+        "$dff"
+          | Just inpst <- minpst -> pure $ \_ -> do
+              cst <- lookupStateFor sc states inpst cnm
+              pure . Just $ Map.fromList
+                [ ("Q", cst)
+                ]
+        _ -> pure $ primCellToMap sc c
+    let
+      -- given a pattern lookup function build a map from output patterns to terms
+      f :: (YosysBitvecConsumer -> Pattern -> m SC.Term) -> m (Map Pattern SC.Term)
+      f = getInps >=> inpsToOuts >=> \case
+        Nothing -> throw $ YosysErrorNoSuchCellType (c ^. cellType) cnm
+        Just outs -> do
+          ms <- forM (Map.toList outs) $ \(onm, otm) ->
+            case Map.lookup onm $ c ^. cellConnections of
+              Nothing -> panic "buildPatternMap" ["Missing expected output name for cell"]
+              Just opat -> deriveTermsByIndices sc opat otm
+          pure $ Map.unions ms
+
+    -- for each output pattern, construct a function that takes a pattern lookup function and computes the associated term
+    fmap Map.fromList . forM derivedOutPatterns $ \pat -> pure . (pat,) $ f >=> \pats -> do
+      case Map.lookup pat pats of
+        Nothing -> panic "buildPatternMap" ["Missing expected output pattern for cell"]
+        Just t -> pure t
+  -- all of the pattern term functions for all of the cells in the module
+  zeroTerm <- liftIO $ SC.scBvConst sc 1 0
+  oneTerm <- liftIO $ SC.scBvConst sc 1 1
+  oneBitType <- liftIO $ SC.scBitvector sc 1
+  xMsg <- liftIO $ SC.scString sc "Attempted to read X bit"
+  xTerm <- liftIO $ SC.scGlobalApply sc (SC.mkIdent SC.preludeName "error") [oneBitType, xMsg]
+  zMsg <- liftIO $ SC.scString sc "Attempted to read Z bit"
+  zTerm <- liftIO $ SC.scGlobalApply sc (SC.mkIdent SC.preludeName "error") [oneBitType, zMsg]
+  pure . Map.unions $ mconcat
+    [ ms
+    , inpTerms
+    , [ Map.fromList
+        [ ( [BitrepZero], const $ pure zeroTerm)
+        , ( [BitrepOne], const $ pure oneTerm )
+        , ( [BitrepX], const $ pure xTerm )
+        , ( [BitrepZ], const $ pure zTerm )
+        ]
+      ]
+    ]
+
+translatePattern ::
+  MonadIO m =>
+  SC.SharedContext ->
+  TranslationContext m ->
+  YosysBitvecConsumer ->
+  Pattern ->
+  m SC.Term
+translatePattern sc ctx c p = do
+  let pmap = ctx ^. translationContextPatternMap
+  case Map.lookup p pmap of
+    Just f -> f $ translatePattern sc ctx
+    Nothing -> do
+      one <- liftIO $ SC.scNat sc 1
+      boolty <- liftIO $ SC.scBoolType sc
+      many <- liftIO . SC.scNat sc . fromIntegral $ length p
+      onety <- liftIO $ SC.scBitvector sc 1
+      bits <- forM p $ \b -> do
+        case Map.lookup [b] pmap of
+          Just t -> t $ translatePattern sc ctx
+          Nothing -> throw $ YosysErrorNoSuchOutputBitvec (Text.pack $ show b) c
+      vecBits <- liftIO $ SC.scVector sc onety bits
+      liftIO $ SC.scJoin sc many one boolty vecBits
+
+translateModule ::
```

Comment:
This function seems to be where a lot of the action happens. As such, I worry that it's under-documented in its current state. I wonder if it is worth including a high-level description of the strategy being used to translate sequential circuits. It looks fairly complex from a glance, involving lots of tricky footwork with internal state.

Reply:
Looks like Sam went through and added a bunch of documentation to `translateModule` (and others).